### PR TITLE
fix(reposerver): loosen source not permitted helm errors

### LINKED ISSUE #13833: Manifest Generation error
Checklist:

* [ ] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [ ] I've included steps to reproduce the bug.
* [ ] I've pasted the output of `argocd version`.

**Describe the bug**

**To Reproduce**

**Expected behavior**

**Screenshots**

**Version**

```shell
Paste the output from `argocd version` here.
```

**Logs**

```
Paste any relevant application logs here.
```

### PR DESCRIPTION
**Pull request by blakepettersson**
With #12255, we check if a source is first permitted before running `helm template`. This works a bit too well, since this may break previously working manifests. If an `AppProject` has a set of `sourceRepos` which are more restrictive than `*`, and it also has Helm public dependencies (repos with credentials would not work with 2.7x due to the fact they get filtered out before ending up on the repo server). Whereas before this would work, this currently fails on `HEAD` but not in `2.7x`.

What we instead do here is that we only run this check if the chart failed to download - if it does then we run a check to see if the repo is in the allowed repos list. If the repo is not in the allowed repos list, we return the same error as in #12555, otherwise we bubble up the error.

Should fix #13833.

Note on DCO:

If the DCO action in the integration test fails, one or more of your commits are not signed off. Please click on the *Details* link next to the DCO action for instructions on how to resolve this.

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] Optional. My organization is added to USERS.md.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)). 
* [ ] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [x] I have added a brief description of why this PR is necessary and/or what this PR solves.

Please see [Contribution FAQs](https://argo-cd.readthedocs.io/en/latest/developer-guide/faq/) if you have questions about your pull-request.

### CODE REVIEW

File: reposerver/repository/repository.go [hunk 1]
```diff
@@ -1023,9 +1023,9 @@ func runHelmBuild(appPath string, h helm.Helm) error {
 	manifestGenerateLock.Lock(appPath)
 	defer manifestGenerateLock.Unlock(appPath)
 
-	// the `helm dependency build` is potentially time consuming 1~2 seconds
-	// marker file is used to check if command already run to avoid running it again unnecessary
-	// file is removed when repository re-initialized (e.g. when another commit is processed)
+	// the `helm dependency build` is potentially a time-consuming 1~2 seconds,
```
Comment (ishitasequeira) [inline reposerver/repository/repository.go:1026]: nit: `a time` -> `time`

File: reposerver/repository/repository.go [hunk 2]
```diff
@@ -1023,9 +1023,9 @@ func runHelmBuild(appPath string, h helm.Helm) error {
...
+	// a marker file is used to check if command already run to avoid running it again unnecessarily
```
Comment (ishitasequeira) [inline reposerver/repository/repository.go:1027]: This might not be related to the changes in this PR, but maybe reword this?

**Review by ishitasequeira — CHANGES_REQUESTED** [review]
Code looks good to me.. The e2e tests do need to be updated!!

**Comment by blakepettersson** [comment]
Not sure what's going on here, when I run `test-e2e` locally the tests pass, but not when running them on CI..

File: reposerver/repository/repository.go (hunk 2 above)
Comment (blakepettersson) [reply reposerver/repository/repository.go:1052]: How would you reword this?

File: reposerver/repository/repository.go (hunk 1 above)
Comment (blakepettersson) [reply reposerver/repository/repository.go:1051]: Hmm, perhaps 

```suggestion
	// the `helm dependency build` potentially takes 1-2 seconds,
```

?

**Comment by blakepettersson** [comment]
So it turns out it was another test that failed than the one I thought was failing 🤦 Now I've updated the test accordingly and the e2e tests now pass.

**Review by ishitasequeira — APPROVED** [review]
Thanks @blakepettersson!! Overall changes look good!! Please commit the suggestion. Just added one nit.

File: reposerver/repository/repository.go (hunk 2 above)
Comment (blakepettersson) [reply reposerver/repository/repository.go:1052]:
```suggestion
	// a marker file is used to check if command already ran to avoid running it again unnecessarily
```
